c10/core/TensorImpl.cpp

```diff
@@ -172,7 +172,16 @@ TensorImpl::TensorImpl(
     // TODO: Ideally we only add AutogradBackend key when the tensor requires
     // grad.
     //       See Note [Dream: skip VariableType kernel when requires_grad=false]
-    key_set_ = key_set | getAutogradRelatedKeySetFromBackend(k);
+
+    // [Note: Nested Tensor Autograd] The Nested Tensor key is a functionality
+    // key and therefore getAutogradRelatedKeySetFromBackend will return the
+    // wrong autograd key For this specific impl we make sure to register the
+    // correct Autograd key
+    if (key_set.has_any(c10::DispatchKeySet{c10::DispatchKey::NestedTensor})) {
+      key_set_ = key_set | inplace_or_view_ks | autograd_nested;
+    } else {
+      key_set_ = key_set | getAutogradRelatedKeySetFromBackend(k);
+    }
```

Comment:
Actually - I think we should add this logic to the constructor for `NestedTensorImpl` instead of adding it directly here. Mostly because:

(1) the custom "NestedTensor -> AutogradNestedTensor" mapping is specific to NT, so it may as well live the the rest of the NT-specific code

(2) This function is very hot-path, and that extra conditional might have a noticeable performance impact on non-NestedTensor code.